### PR TITLE
Added PPM support

### PR DESCRIPTION
**Pull request by edatki**
Added some code to decode a PPM signal from an RC receiver.

### CODE REVIEW

File: Firmware/RoboPad/RoboPad.ino [hunk 1]
```diff
@@ -423,22 +432,64 @@ void setup() {
   // Set up the websocket server:
   webSocketServer.begin();
   webSocketServer.onEvent(webSocketEvent);
+  Serial.println("WebSocket server started.");
 
   // Configure IO
   configureIO();
+
+  attachInterrupt(PPM_PIN, interruptPPM, CHANGE);
+  Serial.println("PPM interrupt attatched");
```
Comment (Blayzeing) [inline Firmware/RoboPad/RoboPad.ino]: These have been missed from the PPM ifdef.

File: Firmware/RoboPad/RoboPad.ino [hunk 2]
```diff
@@ -368,6 +369,14 @@ bool safeboot = false;
 #define RADIO_SILENCE_AUTO_CUTOFF 3000 // If there's no data in 3 seconds then shut down.
 #define SAFEBOOT_PIN 16 // Used as a switch (pulled low by circuit, pull high to safeboot)
 
+#ifdef PPM
+#define PPM_CHANNELS 8
+#define PPM_PIN 3
+#define PPM_MIN_PULSE_DURATION 550
+#define PPM_MAX_PULSE_DURATION 1850
+unsigned char channels[PPM_CHANNELS];
```
Comment (Blayzeing) [inline Firmware/RoboPad/RoboPad.ino]: I feel like this should be volatile.

File: Firmware/RoboPad/RoboPad.ino [hunk 3]
```diff
@@ -423,22 +432,64 @@ void setup() {
...
+
+  attachInterrupt(PPM_PIN, interruptPPM, CHANGE);
+  Serial.println("PPM interrupt attatched");
 }
 
+#ifdef PPM
+long counter = 0;
+long startMicros = 0;
+long minSync = 2500;
+bool ppmActive = false;
```
Comment (Blayzeing) [inline Firmware/RoboPad/RoboPad.ino]: I feel like this should be volatile.

File: Firmware/RoboPad/RoboPad.ino [hunk 4]
```diff
@@ -423,22 +432,64 @@ void setup() {
...
 void loop() {
   httpServer.handleClient();
 
   if(safeboot) return;// Make sure we don't do anything when safeboot is operational.
+
+  #ifdef PPM
+  if (ppmActive) {
+    esc.writeMicroseconds(map(channels[2], 0, 255, 1000, 2000));
+    leftMotor.drive(channels[0], 0, 255);
+    rightMotor.drive(channels[1], 0, 255);
```
Comment (Blayzeing) [inline Firmware/RoboPad/RoboPad.ino]: Haven't had to deal with it in a while, but if `channels` is made volatile then I _think_ this should be wrapped in an ATOMIC_BLOCK.

**Review by Blayzeing — COMMENTED** [review]
Thanks for pushing this! I like it but I've left a few comments that I just want to double-check before merging.

File: Firmware/RoboPad/RoboPad.ino (hunk 1 above)
Comment (edwardatki) [reply Firmware/RoboPad/RoboPad.ino]: Fixed this in latest commit

File: Firmware/RoboPad/RoboPad.ino (hunk 2 above)
Comment (edwardatki) [reply Firmware/RoboPad/RoboPad.ino]: Fixed this in latest commit

File: Firmware/RoboPad/RoboPad.ino (hunk 3 above)
Comment (edwardatki) [reply Firmware/RoboPad/RoboPad.ino]: Fixed this in latest commit

File: Firmware/RoboPad/RoboPad.ino (hunk 4 above)
Comment (edwardatki) [reply Firmware/RoboPad/RoboPad.ino]: I'm not familiar with ATOMIC_BLOCK. If you want to add it feel free but it seems to work fine without it.

Comment (Blayzeing) [reply Firmware/RoboPad/RoboPad.ino]: No probs, I've merged now - thanks for contributing! I'll add an issue to look into ATOMIC_BLOCK use, but if it works then it works :)

**Pull request merged**